plugins/java-decompiler/engine/src/org/jetbrains/java/decompiler/modules/decompiler/exps/ConstExprent.java

```diff
@@ -234,6 +241,138 @@ else if (constType.equals(VarType.VARTYPE_CLASS)) {
+      }
+      else if (doubleVal == Double.POSITIVE_INFINITY) {
+        return new FieldExprent("POSITIVE_INFINITY", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+      }
+      else if (doubleVal == Double.NEGATIVE_INFINITY) {
+        return new FieldExprent("NEGATIVE_INFINITY", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+      }
+      else if (doubleVal == Double.MAX_VALUE) {
+        return new FieldExprent("MAX_VALUE", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+      }
+      else if (doubleVal == Double.MIN_NORMAL) {
+        return new FieldExprent("MIN_NORMAL", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+      }
+      else if (doubleVal == Double.MIN_VALUE) {
+        return new FieldExprent("MIN_VALUE", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+      }
+      else if (doubleVal == -Double.MAX_VALUE) {
+        return new FieldExprent("MAX_VALUE", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer).prepend("-");
+      }
+      else if (doubleVal == -Double.MIN_NORMAL) {
+        return new FieldExprent("MIN_NORMAL", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer).prepend("-");
+      }
+      else if (doubleVal == -Double.MIN_VALUE) {
+        return new FieldExprent("MIN_VALUE", "java/lang/Double", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer).prepend("-");
+      }
+      // Math constants
+      else if (doubleVal == Math.E) {
+        return new FieldExprent("E", "java/lang/Math", true, null, FieldDescriptor.DOUBLE_DESCRIPTOR, bytecode).toJava(0, tracer);
+      }
+      else if (PI_DOUBLES.containsKey(doubleVal)) {
+        String[] parts = PI_DOUBLES.get(doubleVal);
+        return getPiDouble(tracer).enclose(parts[0], parts[1]);
+      }
+
+      // Check for cases where a float literal has been upcasted to a double.
+      // (for instance, double d = .01F results in 0.009999999776482582D without this)
+      float nearestFloatVal = (float)doubleVal;
+      if (doubleVal == (double)nearestFloatVal) {
+        // Value can be represented precisely as both a float and a double.
+        // Now check if the string representation as a float is nicer/shorter.
+        // If they're the same, there's no point in the cast and such (e.g. don't decompile 1.0D as (double)1.0F).
+        if (Float.toString(nearestFloatVal).length() < Double.toString(doubleVal).length()) {
+          // Include a cast to prevent using the wrong method call in ambiguous cases.
+          return createFloat(literal, nearestFloatVal, tracer).prepend("(double)");
+        }
+      }
```

Comment:
Not sure about this. Upscaling floats to doubles doesn't seem too common, and additional casts may harm readability more than shorter literals are able to improve. Personally, I'd prefer to see an exact double value instead of a float approximation - as an additional reminder that IEEE 754 is tricky :)

Reply:
The main reason I do this is because it happens in a surprisingly large number of places in the Minecraft codebase.  I don't know *why* Mojang has so many, but they do, and it makes things a lot more readable for e.g. `9.8F` and `0.1F`.  (See diff linked in the downstream PR).